### PR TITLE
Nodes:  Add LineDashedNodeMaterial

### PR DESCRIPTION
**Pull request by aardgoose**
Add  support for dashed lines.

@sunag 
materialReferences() fail when accessed across code flows. The generate() methods added resolve the problem but there is probably a better fix. The required outputNodes for the lineScale reference are cached against the fragment shader but required in the vertex shader. Without the changes the required uniform name is missing from the generated code.

ie (pseudo code).

uniform lineScale;

varying = ( * lineLength );

instead of:

varying = ( lineScale * lineLength );

### CODE REVIEW

File: examples/jsm/nodes/accessors/LineMaterialNode..js [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+	}
+
+
+	construct( /* builder */ ) {
+
+		const node = this.getFloat( this.scope );
+		this.node = node;
+
+		return node;
```
Comment (sunag) [inline examples/jsm/nodes/accessors/LineMaterialNode..js:21]: The code below should work, or maybe we have some issue in the build process.

```js
class LineMaterialNode extends MaterialNode {

	constructor( scope ) {

		super();

		this.scope = scope;

	}


	construct( /* builder */ ) {

		return this.getFloat( this.scope );

	}

}
```

Comment (aardgoose) [reply examples/jsm/nodes/accessors/LineMaterialNode..js:21]: I tried that initially, but it failed as described. I traced it through the path for varying()  where the reference is included in the vertex shader flow and the uniform name is omitted in the code produced. 

The output from the construct() phase is entered into the nodes properties[ 'fragment' ] so isn't available at that point. I don't understand the code well enough to known if there is a better solution.

Comment (sunag) [reply examples/jsm/nodes/accessors/LineMaterialNode..js:21]: Hmm... Make sense, `construct()` must build the output node once. I will fix this.

File: examples/jsm/nodes/materials/LineDashedNodeMaterial.js [hunk 2]
```diff
@@ -0,0 +1,39 @@
+import NodeMaterial, { addNodeMaterial } from './NodeMaterial.js';
+import { attribute } from '../core/AttributeNode.js';
+import { varying } from '../core/VaryingNode.js';
+import { lineScale, lineDashSize, lineGapSize } from '../accessors/LineMaterialNode..js';
```
Comment (sunag) [inline examples/jsm/nodes/materials/LineDashedNodeMaterial.js:4]: Seems to have two points in `..js`

Comment (aardgoose) [reply examples/jsm/nodes/materials/LineDashedNodeMaterial.js:4]: fixed

File: examples/jsm/nodes/accessors/LineMaterialNode.js [hunk 3]
```diff
@@ -0,0 +1,31 @@
+import MaterialNode from './MaterialNode.js';
+import { addNodeClass } from '../core/Node.js';
+import { nodeImmutable } from '../shadernode/ShaderNode.js';
+
+class LineMaterialNode extends MaterialNode {
+
+	constructor( scope ) {
```
Comment (LeviPesin) [inline examples/jsm/nodes/accessors/LineMaterialNode.js:7]: Do we need such a pass-through constructor? I think these lines (7-11) can be just removed.

File: examples/jsm/nodes/core/PropertyNode.js [hunk 4]
```diff
@@ -57,5 +57,7 @@ export const iridescenceThickness = nodeImmutable( PropertyNode, 'float', 'Iride
 export const specularColor = nodeImmutable( PropertyNode, 'color', 'SpecularColor' );
 export const shininess = nodeImmutable( PropertyNode, 'float', 'Shininess' );
 export const output = nodeImmutable( PropertyNode, 'vec4', 'Output' );
+export const dashSize = nodeImmutable( PropertyNode, 'float', 'dashScale' );
+export const gapSize= nodeImmutable( PropertyNode, 'float', 'gapSize' );
```
Comment (LeviPesin) [inline examples/jsm/nodes/core/PropertyNode.js:61]: Missing space before the `=` sign.

**Pull request merged**